rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,777 @@
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&MQTTTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    pub fn get_tx_by_pkt_id(&mut self, pkt_id: u32) -> Option<&mut MQTTTransaction> {
+        for tx in &mut self.transactions {
+            if let Some(mpktid) = tx.pkt_id {
+                if mpktid == pkt_id {
+                    return Some(tx);
+                }
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self, msg: MQTTMessage, toclient: bool) -> MQTTTransaction {
+        let mut tx = MQTTTransaction::new(msg);
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        if toclient {
+            tx.toclient = true;
+        } else {
+            tx.toserver = true;
+        }
+        return tx;
+    }
+
+    fn handle_msg(&mut self, msg: MQTTMessage, toclient: bool) {
+        match msg.op {
+            MQTTOperation::CONNECT(ref conn) => {
+                self.protocol_version = conn.protocol_version;
+                if let Some(tx) = self.get_tx_by_pkt_id(MQTT_CONNECT_PKT_ID) {
+                    &mut MQTTState::set_event(tx, MQTTEvent::DoubleConnect);
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    tx.pkt_id = Some(MQTT_CONNECT_PKT_ID);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::PUBLISH(ref publish) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                match msg.header.qos_level {
+                    0 => {
+                        // with QOS level 0, we do not need to wait for a
+                        // response
+                        let mut tx = self.new_tx(msg, toclient);
+                        tx.complete = true;
+                        self.transactions.push(tx);
+                    },
+                    1..=2 => {
+
+                        if let Some(pkt_id) = publish.message_id {
+                            let mut tx = self.new_tx(msg, toclient);
+                            tx.pkt_id = Some(pkt_id as u32);
+                            self.transactions.push(tx);
+                        } else {
+                            let mut tx = self.new_tx(msg, toclient);
+                            &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingMsgId);
+                            self.transactions.push(tx);
+                        }
+
+                    },
+                    _ => {
+                        let mut tx = self.new_tx(msg, toclient);
```

Comment:
This line is done for every case

Reply:
I was trying to only get _one_ mutable reference late in the code flow, otherwise I would need to pass around and store references with lifetimes. Trying to keep things simple...
Moving this line before the `match msg...` would require this... right?

Reply:
Can you add this as a comment ? thanks for the justification

Reply:
Done. Maybe there is a more concise way to do it, but this works for now. I'd rather have a clear view of what is happening 
 each case instead of just reducing the line count.